trpcultivate_genotypes/src/GenotypesLoader/GenotypesLoaderPluginBase.php

```diff
@@ -224,6 +242,149 @@ public function getRecordPkey(string $record_type, string $table, int $mode, arr
+    $SAMPLES_FILE = fopen($sample_file, 'r');
+
+    // Grab the header and count the number of columns
+    $header = fgetcsv($SAMPLES_FILE, 0, "\t");
+    $num_columns = count($header);
+    if (!(($num_columns >= 5) && ($num_columns <= 7))) {
+      throw new \Exception(
+        t("Unexpected number of columns (%columns) in the samples file: %file", ['@file'=>$sample_file, '@columns'=>$num_columns])
+      );
+    }
+
+    // Collect our default germplasm type and organism
+    $default_germplasm_type_id = $genetics_config->get('terms.germplasm_type');
+    $default_organism_id = $this->getOrganismID();
+
+    // Iterate through each row to grab all of the samples
+    while(!feof($SAMPLES_FILE)) {
+      $current_line = fgetcsv($SAMPLES_FILE, 0, "\t");
+      if (empty($current_line)) continue;
+
+      // Column 1: DNA source (name should match sample in the input file)
+      $source_name = array_shift($current_line);
+      // Column 2: Name of the sample assayed
+      $sample_name = array_shift($current_line);
+      // Column 3: Accession of the sample assayed
+      $sample_accession = array_shift($current_line);
+      // Column 4: Name of the germplasm
+      $germplasm_name = array_shift($current_line);
+      // Column 5: Accession of the germplasm
+      $germplasm_accession = array_shift($current_line);
+      // Column 6: User can optionally supply a stock_type for each germplasm if they are inserting
+      if ($num_columns >= 6) {
+        $germplasm_type = array_shift($current_line);
+        // Break our germplasm type into its dbname and accession
+        list($germplasm_type_dbname, $germplasm_type_accession) = explode(':', $germplasm_type);
+        // Reminder: check that germplasm_type is in the correct format
+        $query = $this->connection->select('1:cvterm', 'cvt')
+          ->fields('cvt', ['cvterm_id']);
+        // Joins cannot be chained
+        $query->join('1:dbxref', 'dbx', 'dbx.dbxref_id = cvt.dbxref_id');
+        $query->join('1:db', 'db', 'dbx.db_id = db.db_id');
+        $query->condition('db.name', $germplasm_type_dbname)
+          ->condition('dbx.accession', $germplasm_type_accession);
+        $records = $query->execute()->fetchAll();
+        // Check there is only 1 record, otherwise throw an exception
+        $germplasm_type_id = $records[0]->cvterm_id;
+
+        // $germplasm_type_id = $this->getRecordPkey('Germplasm Type', 'cvterm', 0, [
+        //   'name' => $germplasm_type
+        // ]);
+        // Reminder: getRecordPkey() will throw an exception if !$germplasm_type_id
+      } else {
+        // If not provided with a cvterm, grab the default
+        $germplasm_type_id = $default_germplasm_type_id;
+      }
+      // Column 7: User can optionally supply an organism for each germplasm if
+      // they are inserting germplasm into the database. We need to allow spaces
+      // in the genus and species, so use a custom query to check that what the user
```

Comment:
This comment mentions using a custom query but instead it should now say use the Tripal core api function. It would also be good to now mention that this supports infraspecific type and name as well.